### ISSUE TITLE
[CT-1472] [Bug] Snapshotting a Table with a PK duplicate and invalidate_hard_deletes=True can leave the snapshot in a non-runnable state

### ISSUE THREAD

**Issue by davem-bis** [dbt-labs/dbt-core#6215]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

When snapshotting a table with a Primary Key duplicate, the affected rows are inserted into the snapshot with the same `dbt_scd_id` value.

If the duplicate rows are then identified and removed from the source table, subsequent snapshots will fail as the `deletes` CTE for the `snapshot__dbt_tmp` view are non-unique.

The `deletes` CTE currently consists of the following columns:
- The string 'delete'
- `NULL` values for every column in the source data.
- `snapshot_get_time()`
- The `dbt_scd_id` column from the snapshot.

Of these, only `dbt_scd_id` is actually being selected from any source or snapshotted data.

### Expected Behavior

The duplicate key rows within the snapshot are updated with a `dbt_valid_to` value and the snapshot succeeds.

This can be achieved by adding a `DISTINCT` clause to the `deletes` CTE in [helpers.sql](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/include/global_project/macros/materializations/snapshots/helpers.sql).

This does not defend against key duplicates in the source table, only existing key duplicates in the snapshot table that have been removed from the source table.

### Steps To Reproduce

DBT Core 1.2.2, Azure Databricks

1. ```sql
    CREATE TABLE IF NOT EXISTS database_name.source_table (
        KeyColumn INT,
        ValueColumn INT
    )
    ```
1. ```sql
    INSERT INTO database_name.source_table (
        KeyColumn,
        ValueColumn
    )
    VALUES
        (1, 1),
        (1, 2)
    ```
1. ```
    {% snapshot source_table_snapshot %}
    
    {{
        config(
          unique_key='KeyColumn',
          strategy='check',
          check_cols = ['ValueColumn'],
          invalidate_hard_deletes=True
        )
    }}

    SELECT
        *
    FROM
        database_name.source_table

    {% endsnapshot %}
   ```
1. Run the snapshot.
1. ```sql
    TRUNCATE TABLE source_table
    ```
1. Run the snapshot. This will fail as there are multiple rows (from the `deletes` CTE) matching each of the duplicate keys in the snapshot table.
    

### Environment

```markdown
- OS: Ubuntu 20.04, Windows 10
- Python: 3.8
- dbt: 1.0.2, 1.2.2
```

### Which database adapter are you using with dbt?

spark

**Comment by dbeatty10** [comment]
Thanks for reporting this @davem-bis ! Your write-up of the reproduction steps and proposed solution are both thoughtful and helpful.

## TL;DR
I think the ultimate solution is to [ensure the unique key is really unique](https://docs.getdbt.com/docs/build/snapshots#ensure-your-unique-key-is-really-unique) to defend against duplicate keys from ever making in into the snapshot table. 

## Problem
The initial problem you described is when a `dbt snapshot` is executed but `unique_key` is not actually unique within the source data. Thereafter, `dbt snapshot` won't run even after you've fixed the problem in the source.

## Option 1

One option is as you described:
- update dbt-core to deduplicate via `DISTINCT` to allow the hard-deletes to be invalidated

### ✅ Pros
- snapshots work again once the source data doesn't contain duplicates 💪 

### ❌ Cons
- 👉 there can still be duplicates when slicing the snapshot destination table on certain timestamps (i.e., the time interval between the first and second snapshots)
    - as such, downstream users of the snapshot could still be exposed to those duplicates
- `DISTINCT` would add overhead to every snapshot, even though most won't need it
- ~this still might not fix the problem when there is a duplicated key, but one or more unique data elements~ EDIT: if `dbt_scd_id`, `'delete'` , and `snapshot_get_time()` are the only non-null columns within the `deletes` CTE and the latter two are constants, then `DISTINCT` would do the trick.

### Variant

A variant on this idea would be to add a `distinct` configuration option which would default to `false`, but can be set to `true` when repairs are needed.

## Option 2

I'm _guessing_ you can achieve something similar to Option 1 (but I didn't actually try it out):
1. set the `unique_key` to something globally unique (like hashing together your actual primary key with the `current_timestamp`)
1. Run a snapshot which should add an end date to all pre-existing rows, and create a new row for each record in the source table
1. Change the `unique_key` back to the original value for all subsequent snapshots 

This is described at a high-level here: https://github.com/dbt-labs/dbt-core/issues/2642#issuecomment-725662391

### ✅ Pros
- no changes needed to dbt-core
- (presumably) achieves the same goals as described in the issue

### ❌ Cons
- still requires some manual actions (albeit all in dbt code)
- creates an extra copy of the source table within the snapshot table

## Option 3

Another option is to:
- use some (probably manual) process outside of dbt to fix the snapshot destination table

### ✅ Pros
- snapshots work again (assuming the source data no longer contains duplicates)
- no overhead from `DISTINCT`
- the user can control exactly how the fix happens

### ❌ Cons
- requires (probably manual) intervention on the part of the user to recover from the duplicate keys 😢 

## Option X

Is there any automated (and low overhead) way to prevent duplicate keys in the first place? 🤷

(This is one of those cases where true database constraints would be handy, but isn't an option since not all data platforms support constraints.)

Our current recommendation is to [add uniqueness tests](https://docs.getdbt.com/docs/build/snapshots#ensure-your-unique-key-is-really-unique) to the source that is being snapshotted.

An alternative to the uniqueness test would be to always deduplicate somehow (for example, using [`dbt_utils.deduplicate()`](https://github.com/dbt-labs/dbt-utils#deduplicate-source)).

## Conclusion
I'm not in love with any of these options! It's more like which is the most tolerable 😰

If I were an end-user with a manageable number of expected deletes, I think I'd just want the `distinct` behavior so I didn't have to worry about it.

But as a maintainer, I'd prefer to not force upon _all_ users the potential performance penalties that would come with implementing the `distinct` solution. We _could_ make this an opt-in configuration option, but I'd also prefer to not add yet another configuration option (if avoidable). This basically leads to Option 2 or 3 (unless there is some other option we haven't explored yet).

So I think the ultimate way to address this is to really make sure the ensure the unique key is **really** unique
 (e.g., via [adding uniqueness tests](https://docs.getdbt.com/docs/build/snapshots#ensure-your-unique-key-is-really-unique)) to prevent this in the first place.
 
I'm closing this in deference to the Ultimate Solution™ in the TLDR above, but truly happy to discuss further if you or anyone else has further insight!
